viamonvifdiscovery/service.go

```diff
@@ -0,0 +1,134 @@
+// Package viamonvifdiscovery provides the discovery service that wraps ONVIF integration for the viamrtsp module
+package viamonvifdiscovery
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/viam-modules/viamrtsp"
+	"github.com/viam-modules/viamrtsp/viamonvif"
+	"github.com/viam-modules/viamrtsp/viamonvif/device"
+	"go.viam.com/rdk/components/camera"
+	"go.viam.com/rdk/logging"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/services/discovery"
+)
+
+// Model is the model for a rtsp discovery service.
+var Model = viamrtsp.Family.WithModel("onvif")
+
+func init() {
+	resource.RegisterService(
+		discovery.API,
+		Model,
+		resource.Registration[discovery.Service, *Config]{
+			Constructor: newDiscovery,
+		})
+}
+
+// Config is the config for the discovery service.
+type Config struct {
+	Credentials []device.Credentials `json:"credentials"`
+}
+
+// Validate validates the discovery service.
+func (cfg *Config) Validate(_ string) ([]string, error) {
+	// check that all creds have both usernames and passwords set. Note a credential can have both fields empty
+	for _, cred := range cfg.Credentials {
+		if cred.Pass != "" && cred.User == "" {
+			return nil, fmt.Errorf("credential missing username, has password %v", cred.Pass)
+		}
+	}
+	return nil, nil
+}
+
+type rtspDiscovery struct {
+	resource.Named
+	resource.AlwaysRebuild
+	Credentials []device.Credentials
+	logger      logging.Logger
+}
+
+func newDiscovery(_ context.Context, _ resource.Dependencies,
+	conf resource.Config,
+	logger logging.Logger,
+) (discovery.Service, error) {
+	cfg, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return nil, err
+	}
+	dis := &rtspDiscovery{
+		Named: conf.ResourceName().AsNamed(),
+		// place the empty credentials first, so we find the insecure cameras before other cameras.
+		Credentials: cfg.Credentials,
+		logger:      logger,
+	}
+
+	return dis, nil
+}
+
+// Close closes the discovery service.
+func (dis *rtspDiscovery) Close(_ context.Context) error {
+	return nil
+}
+
+// DiscoverResources discovers different rtsp cameras that use onvif.
+func (dis *rtspDiscovery) DiscoverResources(ctx context.Context, _ map[string]any) ([]resource.Config, error) {
+	potentialCams := []resource.Config{}
+
+	list, err := viamonvif.DiscoverCameras(ctx, dis.Credentials, nil, dis.logger)
```

Comment:
currently I am not using any manual urls. Can add these to the config if requested, pending a rescope.